File: backend/pubsub/message.go
package pubsub

import "encoding/json"

type BaseMessage struct {
	Type string          `json:"type"`
	Data json.RawMessage `json:"data"`
}

type SubscribeMessage struct {
	Topic string `json:"topic"`
}

type UserIdMessage struct {
	Type   string `json:"type"`
	UserId string `json:"userId"`
}

type PublishPositionMessage struct {
	Type            string          `json:"type"`
	UserId          string          `json:"userId"`
	PositionMessage PositionMessage `json:"data"`
}

type PublishPlayerExitMessage struct {
	Type   string `json:"type"`
	UserId string `json:"userId"`
}

type PositionMessage struct {
	X int `json:"x"`
	Y int `json:"y"`
}

File: frontend/src/models/message.ts
interface BaseMessage {
  type: string;
}

interface UserIdMessage extends BaseMessage {
  type: "userId";
  userId: string;
}

interface PositionMessage extends BaseMessage {
  type: "position";
  data: {
    x: number;
    y: number;
  };
}

interface PublishPositionMessage extends PositionMessage {
  userId: string;
}

interface PublishPlayerEntranceMessage extends BaseMessage {
  type: "playerEntrance",
  userId: string
}

interface PublishPlayerExitMessage extends BaseMessage {
  type: "playerExit",
  userId: string
}

export type incomingMessage = UserIdMessage | PublishPositionMessage | PublishPlayerExitMessage | PublishPlayerEntranceMessage;
export type outgoingMessage = PositionMessage;
